### ISSUE TITLE
JEMalloc leaks mapped pages on Windows

### ISSUE THREAD

**Issue by mjp41** [jemalloc/jemalloc#487]
Using 4.2.1 we found the following program uses aproximately 3Gb of memory, rather than less than 1 Mb.

```C
#define JEMALLOC_EXPORT
#include "jemalloc.h"

#define DELAY 12
#define CHUNK 3

void 
main() 
{
    void ** buffer = (void**)je_calloc(sizeof(void*),  DELAY);

    for(int i = 0; i< 4000000; i++) {
        if(i % CHUNK == 0) {
            for (int k = 0; k < CHUNK; k++) {
                if (buffer[(i + k) % (DELAY)] != 0)
                    je_free(buffer[(i + k) % (DELAY)]);
            }
        }

        buffer[i % (DELAY)] = je_malloc(40000);
    }
    je_malloc_stats_print(NULL, NULL, NULL);
}
```
The stats print gives
```
___ Begin jemalloc statistics ___
Version: 0.0.0-0-g0000000000000000000000000000000000000000
Assertions disabled
config.malloc_conf: ""
Run-time option settings:
  opt.abort: false
  opt.lg_chunk: 16
  opt.dss: "secondary"
  opt.narenas: 32
  opt.purge: "ratio"
  opt.lg_dirty_mult: 3 (arenas.lg_dirty_mult: 3)
  opt.stats_print: false
  opt.junk: "false"
  opt.quarantine: 0
  opt.redzone: false
  opt.zero: false
  opt.tcache: true
  opt.lg_tcache_max: 15
CPUs: 8
Arenas: 32
Pointer size: 8
Quantum size: 16
Page size: 4096
Min active:dirty page ratio per arena: 8:1
Maximum thread-cached size class: 32768
Chunk size: 65536 (2^16)
Allocated: 724096, active: 794624, metadata: 3284535616, resident: 3285082112, mapped: 52437811200, retained: 0
Current active ceiling: 851968

arenas[0]:
assigned threads: 1
dss allocation precedence: disabled
min active:dirty page ratio: 8:1
purging: dirty: 0, sweeps: 4533322, madvises: 5066652, purged: 61066487
                            allocated      nmalloc      ndalloc    nrequests
small:                         187520          169            0           69
large:                         438272      4000001      3999990      4000001
huge:                           98304            1            0            1
total:                         724096      4000171      3999990      4000071
active:                        794624
mapped:                   52429488128
retained:                           0
metadata: mapped: 3276836864, allocated: 304896
bins:           size ind    allocated      nmalloc      ndalloc    nrequests      curregs      curruns regs pgs  util       nfills
 nflushes      newruns       reruns
                     ---
                  96   6         9600          100            0            0          100            1  128   3 0.781            1
        0            1            0
                     ---
                 128   8          256            2            0            2            2            1   32   1 0.062            0
        0            1            0
                     ---
                 256  12         8704           34            0           34           34            3   16   1 0.708            0
        0            3            0
                     ---
                5120  29       168960           33            0           33           33            9    4   5 0.916            0
        0            9            0
                     ---
large:          size ind    allocated      nmalloc      ndalloc    nrequests      curruns
                     ---
               28672  39        28672            1            0            1            1
                     ---
               40960  41       409600      4000000      3999990      4000000           10
                     ---
huge:           size ind    allocated      nmalloc      ndalloc    nrequests   curhchunks
                     ---
               98304  46        98304            1            0            1            1
                     ---
--- End jemalloc statistics ---
```

This is on Windows using Visual Studio 2015
```
Microsoft (R) C/C++ Optimizing Compiler Version 19.00.23918 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

test.c
Microsoft (R) Incremental Linker Version 14.00.23918.0
Copyright (C) Microsoft Corporation.  All rights reserved.
```
Linked against the static release mode library of JEMalloc.

Interestingly, without returning in chunks (i.e. three at a time), then it performs perfectly.  

I have not managed to build rel-4.3.0 or dev on my machine in Windows yet.

**Comment by mjp41** [comment]
Further information, I experience the same behaviour with rel-4.3.0 (1ceae2f8cbd8).

On dev (04e1328ef1), it crashes with 
```
<jemalloc>: Memory allocation failure in mallctl("epoch", ...)
```

**Comment by jasone** [comment]
Thank you for the test program.  It uncovered two bugs in rel-4.3.0:
- e9012630acf897ce7016e427354bb46fbe893fe1 is Windows-specific (decommit-related), and the bug was responsible for permanently fragmenting chunks.
- 32896a902bb962a06261d81c9be22e16210692db is a generic fix (high impact unless ```--disable-cache-oblivious``` was specified).

On the dev branch, there were related but not identical issues, which are fixed in 8dd5ea87cac39d9a90dbe40d13267ec02df0214c, ea9961acdbc9f7e2c95a3b55ce0ac1024af5d167, and 4a7852137d8b6598fdb90ea8e1fd3bc8a8b94a3a.

**Comment by mjp41** [comment]
Thanks @jasone. Great work.
